Lazily import many modules to improve import speed

ashb opened the pull request:
Goal here is to make the healthchecks for triggerer/scheduler quicker by having to do less

Roughly:

- Don't import jsonschema (module _or_ schemas) until we need it - it was surprisingly slow.
- Dont import all models when importing any one -- the only times we need that are when auto-generating migrations, or when checking DB against schema. (That said, accesing any class will cause the SQLA event to import all models, otherwise we often end up with issues due to the relationships on classes)
- Move a few more things to be lazy/delayed import too.

ashb committed:
WIP: Lazy import and do less at import time

This may be "overkill" for the benefit it actually gives.

mik-laj commented on airflow/jobs/__init__.py:
```diff
@@ -16,8 +16,3 @@
 # specific language governing permissions and limitations
 # under the License.
 #
-import airflow.jobs.backfill_job
```
It is a bit risky. See: https://github.com/apache/airflow/pull/19185

ashb replied:
@mik-laj Yeah I had a feeling something like that might happen - hence why this is in draft and I've got the "load_all_models" fn in airflow.models

ashb replied:
Found a good solution to this that I think is foolproof https://github.com/apache/airflow/pull/24486/commits/685498fa0800c3ca0e51ac41cbd3e863bdab4da6

ashb committed:
fixup! WIP: Lazy import and do less at import time

ashb committed:
fixup! WIP: Lazy import and do less at import time

ashb committed:
fixup! WIP: Lazy import and do less at import time

ashb committed:
fixup! WIP: Lazy import and do less at import time

uranusjr committed:
Sort imports

uranusjr commented on airflow/serialization/json_schema.py:
```diff
@@ -19,14 +19,15 @@
 from airflow.settings import json
 from airflow.typing_compat import Protocol
 
+if TYPE_CHECKING:
+    import jsonschema
```
Does this change much?

ashb replied:
It was high up in the list of slow imports!

uranusjr replied:
A schema validation library being slow… I guess that’s how third-party libraries work.

ashb committed:
Use sql-alchemy events to import_all_models at the "right" time

ashb committed:
Move more "slow" imports out of top level

ashb committed:
Hacky method to let mypy know about lazy imports in airflow.models

For some reason I couldn't get the typestub working: if I created
`airflow/models/__init__.pyi` then Mypy couldn't find any of the
submodules

ashb committed:
fixup! WIP: Lazy import and do less at import time

ashb committed:
Don't force loading of all models early in CLI audit logging

uranusjr commented on airflow/utils/cli.py:
```diff
@@ -156,17 +153,6 @@ def _build_metrics(func_name, namespace):
-        dag_id=metrics.get('dag_id'),
-        execution_date=metrics.get('execution_date'),
-    )
-    metrics['log'] = log
```
So this metric is lost forever…? (although I don’t really understand how this works)

ashb replied:
It was only used to write the Log object elsewhere. So I just created it there instead.

("metric" was an odd name for this variable in the first place)

uranusjr commented on airflow/models/__init__.py:
```diff
@@ -78,3 +50,94 @@
+
+STATICA_HACK = True
+globals()['kcah_acitats'[::-1].upper()] = False
+if STATICA_HACK:  # pragma: no cover
```
Would `TYPE_CHECKING` not work here?

ashb replied:
oh yes, probably! Testing

ashb replied:
Yup this works.

ashb committed:
less hacky method

potiuk reviewed (commented):
Nice one :)
